File: 1975.maximum-matrix-sum.cpp
/*
 * @lc app=leetcode id=1975 lang=cpp
 *
 * [1975] Maximum Matrix Sum
 */

#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution {
 public:
  long long maxMatrixSum(vector<vector<int>>& matrix) {
    long long sum = 0;
    int n_minus = 0;
    int min_abs = 100001;
    for (auto row : matrix) {
      for (auto elem : row) {
        sum += abs(elem);
        min_abs = min(min_abs, abs(elem));
        if (elem < 0) {
          ++n_minus;
        }
      }
    }

    if (n_minus % 2 == 0)
      return sum;
    else
      return sum - 2 * min_abs;
  }
};
// @lc code=end

File: 1975.maximum-matrix-sum.rs
/*
 * @lc app=leetcode id=1975 lang=rust
 *
 * [1975] Maximum Matrix Sum
 */

// @lc code=start
impl Solution {
    pub fn max_matrix_sum(matrix: Vec<Vec<i32>>) -> i64 {
        let mut sum: i64 = 0;
        let mut n_minus = 0;
        let mut min_abs = 100001;

        for row in matrix.iter() {
            for &elem in row.iter() {
                sum += elem.abs() as i64;
                min_abs = std::cmp::min(min_abs, elem.abs());

                if (elem < 0) {
                    n_minus += 1;
                }
            }
        }

        if (n_minus % 2 == 0) {
            sum
        } else {
            sum - 2 * (min_abs as i64)
        }
    }
}
// @lc code=end
